File: src/org/cappuccino_project/ide/intellij/plugin/indices/ObjJParentClassChildrenIndex.java
package org.cappuccino_project.ide.intellij.plugin.indices;

import com.intellij.openapi.progress.ProgressIndicatorProvider;
import com.intellij.openapi.project.Project;
import com.intellij.psi.search.GlobalSearchScope;
import com.intellij.psi.stubs.StringStubIndexExtension;
import com.intellij.psi.stubs.StubIndex;
import com.intellij.psi.stubs.StubIndexKey;
import org.cappuccino_project.ide.intellij.plugin.psi.interfaces.ObjJClassDeclarationElement;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.List;

public class ObjJParentClassChildrenIndex extends StringStubIndexExtension<ObjJClassDeclarationElement> {
    private static final ObjJParentClassChildrenIndex INSTANCE = new ObjJParentClassChildrenIndex();
    private static final StubIndexKey<String, ObjJClassDeclarationElement> KEY = IndexKeyUtil.createIndexKey(ObjJParentClassChildrenIndex.class);
    private static final int VERSION = 1;
    private ObjJParentClassChildrenIndex() {}

    public static ObjJParentClassChildrenIndex getInstance() {
        return INSTANCE;
    }

    @Override
    public int getVersion() {
        return super.getVersion()+ ObjJIndexService.INDEX_VERSION+VERSION;
    }

    @NotNull
    @Override
    public StubIndexKey<String, ObjJClassDeclarationElement> getKey() {
        return KEY;
    }

    @NotNull
    public List<String> getChildClassesAsStrings(@NotNull String parentClassName, @NotNull Project project) {
        //ProgressIndicatorProvider.checkCanceled();
        List<String> childClasses = new ArrayList<>();
        List<ObjJClassDeclarationElement> childClassElements = getChildClasses(parentClassName, project);
        for (ObjJClassDeclarationElement childClass : childClassElements) {
            //ProgressIndicatorProvider.checkCanceled();
            String childClassName = childClass.getStub() != null ? childClass.getStub().getClassName() : childClass.getClassNameString();
            if (childClassName.isEmpty()) {
                continue;
            }
            childClasses.add(childClassName);
        }
        return childClasses;
    }

    @NotNull
    public List<ObjJClassDeclarationElement> getChildClasses(@NotNull String parentClassName, @NotNull Project project) {
        return getChildClasses(parentClassName, project, null);
    }

    @NotNull
    public List<ObjJClassDeclarationElement> getChildClasses(@NotNull String parentClassName, @NotNull Project project, @Nullable GlobalSearchScope scope) {
        ProgressIndicatorProvider.checkCanceled();
        return new ArrayList<>(StubIndex.getElements(KEY, parentClassName, project, scope, ObjJClassDeclarationElement.class));
    }
}

File: src/org/cappuccino_project/ide/intellij/plugin/indices/ObjJParentClassChildrenIndex.kt
package org.cappuccino_project.ide.intellij.plugin.indices

import com.intellij.openapi.progress.ProgressIndicatorProvider
import com.intellij.openapi.project.Project
import com.intellij.psi.search.GlobalSearchScope
import com.intellij.psi.stubs.StringStubIndexExtension
import com.intellij.psi.stubs.StubIndex
import com.intellij.psi.stubs.StubIndexKey
import org.cappuccino_project.ide.intellij.plugin.psi.interfaces.ObjJClassDeclarationElement

import java.util.ArrayList

class ObjJParentClassChildrenIndex private constructor() : StringStubIndexExtension<ObjJClassDeclarationElement<*>>() {

    override fun getVersion(): Int {
        return super.getVersion() + ObjJIndexService.INDEX_VERSION + VERSION
    }

    override fun getKey(): StubIndexKey<String, ObjJClassDeclarationElement<*>> {
        return KEY
    }

    fun getChildClassesAsStrings(parentClassName: String, project: Project): List<String> {
        //ProgressIndicatorProvider.checkCanceled();
        val childClasses = ArrayList<String>()
        val childClassElements = getChildClasses(parentClassName, project)
        for (childClass in childClassElements) {
            //ProgressIndicatorProvider.checkCanceled();
            val childClassName = if (childClass.stub != null) childClass.stub!!.className else childClass.classNameString
            if (childClassName.isEmpty()) {
                continue
            }
            childClasses.add(childClassName)
        }
        return childClasses
    }

    @JvmOverloads
    fun getChildClasses(parentClassName: String, project: Project, scope: GlobalSearchScope? = null): List<ObjJClassDeclarationElement<*>> {
        ProgressIndicatorProvider.checkCanceled()
        return ArrayList(StubIndex.getElements<String, ObjJClassDeclarationElement<*>>(KEY, parentClassName, project, scope, ObjJClassDeclarationElement::class.java))
    }

    companion object {
        val instance = ObjJParentClassChildrenIndex()
        private val KEY = IndexKeyUtil.createIndexKey<String, ObjJClassDeclarationElement<*>>(ObjJParentClassChildrenIndex::class.java)
        private val VERSION = 1
    }
}
